### ISSUE TITLE
Longest Flow Path Produces Incorrect Results

### ISSUE THREAD

**Issue by DJHostetler** [jblindsay/whitebox-tools#289]
When using the [LongestFlowpath](https://www.whiteboxgeo.com/manual/wbt_book/available_tools/hydrological_analysis.html#longestflowpath) tool with a subbasins raster (multiple subbasins), the output does not produce the longest flow path for each individual subbasin. If there is no upstream subbasin, it calculates the longest flowpath correctly, but otherwise the results are not the longest flow path for the local subbasin.

**Comment by jfbourdon** [comment]
I recently used the tool and me too I was puzzled by the result that didn't match what I was expecting i.e. the longest flowpath for each basin. The documentation says *If subbasins are input to LongestFlowpath, each traced flowpath will include only the non-overlapping portions within nested areas.* but it doesn't mean what we think.

I looked at the code and it first look all "head" basins (will no upstream other basin) and for those we will get the longest flowpath. After that, the outlet of each of those basins is simply followed downstream until the end of the stream. It means that the flowpaths extracted for any downstream basin isn't guaranteed to be the longest, only that it is connected to a longest flowpath situated in a head basin.

So it doesn't seem to be a bug but rather a feature. That being said, there is the limitation that this tool really expect the basins to come from Subbasins or Wathershed as using Isobasins as input basins will result in a bunch of disconnected flowpaths.

Here is an example with `Subbasins` with pixels representing each basin, the arrows the flow direction, the blue dots the outlets, the green lines the result of `LongestFlowpath` and the red lines *missing* longer local flowpath.


```python
from whitebox_tools import WhiteboxTools
wbt = WhiteboxTools()

wdir = R"C:\Temp\test_LongestFlowpath"
wbt.set_working_dir(wdir)


# Make D8 pointer
dem = "dem.flt"   # -->  https://github.com/jblindsay/whitebox-tools/files/6919277/dem.zip
d8pntr = "d8pntr.flt"
wbt.d8_pointer(dem, d8pntr)

# Make D8 flow accumulation
d8accum = "d8accum.flt"
wbt.d8_flow_accumulation(dem, d8accum)

# Reclass flow accumulation to create distinct streams
streams = "streams.flt"
reclass_vals = ";".join(["-32768;1;10", "1;10;20", "2;20;55", "3;55;75"])
wbt.reclass(d8accum, streams, reclass_vals)

# Make subbasins
subbsins = "subbasins.flt"
wbt.subbasins(d8pntr, streams, subbsins)

# Find longest flowpaths
flowpaths = "flowpaths_subbasins.shp"
wbt.longest_flowpath(dem, subbsins, flowpaths)
```

![image](https://user-images.githubusercontent.com/32580398/211410748-a5883c9f-5ad0-4b2f-88f8-ed9fd2742a45.png)


Doing the same thing with `Isobasins` gives this:

```python
isobasins = "isobasins.flt"
wbt.isobasins(dem, isobasins, 10)

flowpaths = "flowpaths_isobasins.shp"
wbt.longest_flowpath(dem, isobasins, flowpaths)
```

![image](https://user-images.githubusercontent.com/32580398/211412865-ceaf5ce1-2a62-44b0-b860-1b563401170c.png)

**Comment by DJHostetler** [comment]
@jfbourdon I agree with your assessment. I ended up coding my own solution for this in Python based on the subbasins and flow direction grids for my application,, but it is not as performant as this tool would be if it were fixed to cover a more typical use case.

**Comment by XPR23** [comment]
Hello. LongestFlowpath still has the problem. I share two images; the first with LongestFlowpath obtained with whitebox, which is incorrect, and the second was obtained with r.accumulate from GRAS GIS, which is correct.

<img width="677" alt="Imagen1" src="https://github.com/jblindsay/whitebox-tools/assets/145178648/218b94a6-8973-47d6-91b8-dd8ae048eadf">
<img width="680" alt="Imagen2" src="https://github.com/jblindsay/whitebox-tools/assets/145178648/c24b978f-6637-446a-a7b8-16d989147e6d">
